src/OpenTelemetry/Metrics/IDataPoint.cs

```diff
@@ -0,0 +1,37 @@
+// <copyright file="IDataPoint.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Metrics
+{
+    public interface IDataPoint
+    {
+        public DateTimeOffset Timestamp { get; }
+
+        public KeyValuePair<string, object>[] Tags { get; }
+
+        public string ValueAsString { get; }
+
+        public void ResetTags(KeyValuePair<string, object>[] tags);
```

Comment:
Could we make the API so that IDataPoint is immutable? Operations like ResetTags or Reset could return a new instance? That would allow DataPoint to be `readonly struct`, which would be better for perf.

Reply:
I would love to find a better/cleaner solution for this.

The current reason for a mutable DataPoint is to avoid an alloc in the hot path. Thus, I'm using a ThreadStatic, and thus, forcing me to "Reset" this object for reuse.

Reply:
I refactor to remove Reset() from public interface. And the Reset() function is now internal and used only by ThreadStaticStorage. And added more comments so people are fully aware of this "less than ideal" design.

Reply:
Resolving for now. Please reopen as necessary.